Mathlib/Tactic/Linter/UnusedAssumptionInType.lean

```diff
@@ -0,0 +1,114 @@
+  if !(← isProp type) then
+    if let some e := declInfo.value? then
+      impossibleArgs ← lambdaTelescope e fun args e ↦ do
+        let e ← e.eraseProofs
+        return impossibleArgs.filter fun (k, _) ↦
+          k < args.size && !e.containsFVar args[k]!.fvarId!
+  if impossibleArgs.isEmpty then return none
+  return some <| .joinSep (impossibleArgs.toList.map Prod.snd) ", "
+
+/--
+Linter that checks for theorems that assume `[Decidable p]`
+but don't use this assumption in the type.
+-/
+@[env_linter] def decidableClassical : Linter where
+  noErrorsFound := "No uses of `Decidable` arguments should be replaced with `classical`"
+  errorsFound := "USES OF `Decidable` SHOULD BE REPLACED WITH `classical` IN THE PROOF."
+  test declName := do
+    if (← isAutoDecl declName) then return none
+    else if Name.isPrefixOf `Decidable declName then return none
+    let names := #[`Decidable, `DecidableEq, `DecidablePred]
+    return ← checkUnusedAssumptionInType (← getConstInfo declName) names
+
+/--
+Linter that checks for theorems that assume `[Inhabited p]`
+but don't use this assumption in the type.
+-/
+@[env_linter] def inhabitedNonempty : Linter where
+  noErrorsFound := "No uses of `Inhabited` arguments should be replaced"
+  errorsFound := "USES OF `Inhabited` SHOULD BE REPLACED WITH `Nonempty` (OR REMOVED)."
+  test declName := do
+    if (← isAutoDecl declName) then return none
+    return ← checkUnusedAssumptionInType (← getConstInfo declName) #[`Inhabited]
+
+/--
+Linter that checks for theorems that assume `[Fintype p]`,
+but don't use this assumption in the type.
+(Instead, `Finite p` can suffice, or the assumption can be fully removed.)
+-/
+@[env_linter] def finiteFintype : Linter where
+  noErrorsFound := "No uses of `Fintype` arguments should be replaced"
+  errorsFound := "USES OF `Fintype` SHOULD BE REPLACED WITH `Finite` (OR REMOVED)."
+  test declName := do
+    if (← isAutoDecl declName) then return none
+    return ← checkUnusedAssumptionInType (← getConstInfo declName) #[`Fintype]
+
+/--
+Linter that checks for theorems that assume `[Encodable p]`,
+but don't use this assumption in the type.
+(Instead, `Countable p` can suffice, or the assumption can be fully removed.)
+-/
+@[env_linter] def encodableCountable : Linter where
+  noErrorsFound := "No uses of `Encodable` arguments should be replaced"
+  errorsFound := "USES OF `Encodable` SHOULD BE REPLACED WITH `Countable` (OR REMOVED)."
+  test declName := do
+    if (← isAutoDecl declName) then return none
+    else if Name.isPrefixOf `Encodable declName then return none
+    return ← checkUnusedAssumptionInType (← getConstInfo declName) #[`Encodable]
+
+/--
+Linter that checks for theorems that assume `[MetricSpace X]` or `[PseudoMetricSpace X]`,
+but don't use this assumption in the type.
+(Instead, `(Pseudo)MetrisableSpace X` can suffice, or the assumption can be fully removed.)
+-/
+@[env_linter] def metricMetrisable : Linter where
```

Comment:
Note that we spell it as `Metrizable` in [TopologicalSpace.MetrizableSpace](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Metrizable/Basic.html#TopologicalSpace.MetrizableSpace)

Reply:
My guess is that this linter doesn't catch anything, because normally you **use** data in `[MetricSpace X]`, otherwise this would be an unused argument. The linter should catch theorems that only use the projection to `UniformSpace`. Another question is how to catch uses of `[UniformSpace X]` that only use projection to `TopologicalSpace X` (should be replaced by `[TopologicalSpace X] [CompletelyRegularSpace X]` once we redefine the latter to avoid dependency on `Real`) or `[UniformSpace X] [Filter.IsCountablyGenerated (uniformity X)]` that only use the projection to `TopologicalSpace X` everywhere else (should be replaced by `[PseudoMetrizableSpace X]`.